Update Angular to support TS 3.9

petebacondarwin opened the pull request:
This is a big PR with lots of commits. But most are in the compiler or are just tweaks to tests. So hopefully it will not be too arduous to review. It will be easier to go commit by commit when reviewing.

bbarry commented on packages/compiler-cli/ngcc/index.ts:
```diff
@@ -8,16 +8,16 @@
-export function process(options: NgccOptions): void|Promise<void> {
+export function process<T extends AsyncNgccOptions|SyncNgccOptions>(options: T):
+    T extends AsyncNgccOptions ? Promise<void>: void;
+export function process(options: AsyncNgccOptions|SyncNgccOptions): void|Promise<void> {
```
Is NgccOptions being deprecated in favor of the underlying union type?

And why have an overload at all?

I'm trying to understand what the purpose of this change is.

petebacondarwin replied:
TS 3.9 does not like the previous approach to typing this function. This is our solution.

petebacondarwin replied:
Oh and the reason for the overload is so that the return type can be specific to the parameters passed in. Note that when `async: true` the return is a `Promise<void>`. When it is `async: false` the return type is just `void`.

alan-agius4 commented on integration/BUILD.bazel:
```diff
@@ -80,6 +80,25 @@ INTEGRATION_TESTS = {
     "service-worker-schema": {},
     "side-effects": {"tags": ["no-ivy-aot"]},
     "terser": {},
+    "typings_test_ts36": {
```
`typings_test_ts36` and `typings_test_ts37` should be removed.

alan-agius4 committed:
build: pin @microsoft/api-extractor to 7.7.11

Version 7.7.12 generates aliases for objects named the same as global symbols which breaks ts-api-guardian since it doesn't support aliases.

See: https://github.com/microsoft/rushstack/issues/1830

Commit:
test(language-service): external template update should not invalidate program

In TypeScript 3.9, the compiler is able to re-use (i.e. not invalidate)
the previous program if only external templates (i.e. no TS files) have
changed.

Commit:
fix(language-service): use empty statement as parent of type node

In TypeScript 3.9, type nodes need to exist in the context of a statement.
This commit ensures that the synthetic type node has such a parent.

petebacondarwin committed:
refactor(compiler-cli): simplify and clarify `TypeScriptReflectionHost.isClass()`

The comment in this function confused me, so I updated it to clarify that
`isClass()` is not true for un-named classes.

Also, I took the opportunity to use a helper method to simplify the function
itself.

alan-agius4 committed:
build: prepare for TypeScript 3.9

- Fix several compilation errors
- Update @microsoft/api-extractor to be compatible with TypeScript 3.9

petebacondarwin committed:
refactor(compiler-cli): delegate `hasBaseClass()` to `getBaseClassExpression()`

The previous implementations of `hasBaseClass()` are almost
identical to the implementation of `getBaseClassExpression()`.
There is little benefit in duplicating this code so this refactoring
changes `hasBaseClass()` to just call `getBaseClassExpression()`.
This allows the various hosts that implement this to be simplified.

petebacondarwin committed:
refactor(ngcc): simplify the `detectKnownDeclaration()` signature

A number of overloads were added to `detectKnownDeclaration()` to
allow it to support `null` being passed through. In practice this could
easily be avoided, which allows the overloads to be removed and the
method signature and implementations to be simplified.

petebacondarwin committed:
refactor(ngcc): remove unused import

The `Import` import from `src/ngtsc/reflection` is not being used.
This commit simply removes this import from the code.

petebacondarwin committed:
test(ngcc): move specs to correct describe block

The recent tests for Enum handling were added to the
incorrect describe block. This commit moves them back
to the correct block.

petebacondarwin committed:
test(compiler-cli): hande TS 3.9 format in emisson tests

In TypeScript 3.9 the emitted JS code has some differences.
This commit updates the tests to be resilient to these changes.

petebacondarwin committed:
refactor(ngcc): ensure unlocker process works in mock file-systems

Previously the path to the unlocker process was being resolved by the
current file-system. In the case that this was a `MockFileSystemWindows`
on a non-Windows operating system, this resulted in an incorrect path
to the entry-point.

Now the path to the entry-point is hand-crafted to avoid being broken by
whatever FileSystem is in use.

petebacondarwin committed:
fix(ngcc): ensure rendering formatters work with IIFE wrapped classes

After the refactoring of the reflection hosts to accommodate
ES2015 classes wrapped in IIFEs. The same treatment needs to
be applied to the rendering formatters.

petebacondarwin committed:
fix(ngcc): ensure reflection hosts can handle TS 3.9 IIFE wrapped classes

In TS 3.9, ES2015 output can contain ES classes that are wrapped in an
IIFE. So now ES2015 class declarations can look like one of:

```
class OuterClass1 {}
```

```
let OuterClass = class InnerClass {};
```

```
var AliasClass;
let OuterClass = AliasClass = class InnerClass {};
```

```
let OuterClass = (() => class InnerClass {}};
```

```
var AliasClass;
let OuterClass = AliasClass = (() => class InnerClass {})();
```

```
let OuterClass = (() => {
  let AdjacentClass = class InnerClass {};
  // ... static properties or decorators attached to `AdjacentClass`
  return AdjacentClass;
})();
```

```
var AliasClass;
let OuterClass = AliasClass = (() => {
  let AdjacentClass = class InnerClass {};
  // ... static properties or decorators attached to `AdjacentClass`
  return AdjacentClass;
})();
```

The `Esm5ReflectionHost` already handles slightly different IIFE wrappers
around function-based classes. This can be substantially reused when
fixing `Esm2015ReflectionHost`, since there is a lot of commonality
between the two.

This commit moves code from the `Esm5ReflectionHost` into the `Esm2015ReflectionHost`
and looks to share as much as possible between the two hosts.

petebacondarwin committed:
docs(ngcc): tidy up typos etc in comments

This file had a few small typos and other issues that have
now been fixed in this commit.............................

petebacondarwin committed:
fix(ngcc): support `defineProperty()` re-exports in CommonJS and UMD

In TypeScript 3.9 some re-export syntaxes have changed to be getter
functions (created by calls to `Object.defineProperty()`) rather than
simple property accessors.

This commit adds support into the CommonJS and UMD reflection hosts
for this style of re-export syntax.

petebacondarwin committed:
refactor(ngcc): rename `ReexportStatement` to `WildcardReexportStatement`

The term `ReexportStatement` is too general for this particular concept.
Here the re-export actually refers to a wildcard where all the module
exports are being re-exported.

When we introduce other re-export statement types later this will be
confusing.

petebacondarwin committed:
fix(ngcc): `viaModule` should be `null` for local imports

In the CommonJS and UMD reflection hosts, the logic for computing the
`viaModule` property of `Declaration` objects was not correct for some
cases when getting the exports of modules.

In these cases it was setting `viaModule` to the path of the local module
rather than `null`.

petebacondarwin committed:
test(docs-infra): update payload sizes after TS 3.9 update

Here are the significant changes found in
`dist/main.js` for AIO...

```
t.\u0275prov = r.Kb({token: t, factory: t.\u0275fac}),
```
to
```
t.\u0275prov = r.Jb({
  token: t,
  factory: function(e) {
    return t.\u0275fac(e)
  }
}),
```

```
function hs(t){const e=ms();
```

to

```
  function fs(t){if(!(t=String(t).trim()))return'';
  const e=t.match(ds);
  return e&&Un(e[1])===e[1]||t.match(hs)&&function(t){let e=!0,n=!0;
  for(let s=0;
  s<t.length;
  s++){const r=t.charAt(s);
  '\''===r&&n?e=!e:'"'===r&&e&&(n=!n)}return e&&n}(t)?t:(Hn()&&console.warn(`WARNING: sanitizing unsafe style value ${t} (see http://g.co/ng/security#xss).`),'unsafe')}function ps(t){const e=ys();
```

```
b=Kt(p[1],0),
e&&(b.projection=e.map(t=>Array.from(t))),
```

to

```
if(b=Kt(f,0),void 0!==e){const t=b.projection=[];
for(let n=0; n<this.ngContentSelectors.length; n++){
  const s=e[n];
  t.push(null!=s?Array.from(s):null)
}
```

```
return o.observedAttributes=Object.keys(i),
n.map(({propName:t})=>t)
    .forEach(t=>{
      Object.defineProperty(o.prototype,t,{ ... })
}),o}
```

to

```
return t.observedAttributes=Object.keys(i),t})();
return n.map(({propName:t})=>t)
    .forEach(t=>{
      Object.defineProperty(o.prototype,t,{ ... })
    }),o}
```

petebacondarwin committed:
test(ngcc): reformat some subject code for tests

Using backtick multiline strings leads to confusing layout
that does not fit with the surrounding indentation. Also it
can lead to test fragility due to automated code formatting.

This commit changes just one set of subject code to use
a more resilient string concatenation approach.

petebacondarwin committed:
fix(compiler-cli): compute the correct target output for `$localize` messages

In some versions of TypeScript, the transformation of synthetic
`$localize` tagged template literals is broken.
See https://github.com/microsoft/TypeScript/issues/38485

We now compute what the expected final output target of the
compilation will be so that we can generate ES5 compliant
`$localize` calls instead of relying upon TS to do the downleveling
for us.

This is a workaround for the TS compiler bug, which could be removed
when this is fixed. But since it only affects ES5 targeted compilations,
which is now not the norm, it has limited impact on the majority of
Angular projects. So this fix can probably be left in indefinitely.

alan-agius4 committed:
test: update payload sizes after TS 3.9 update

With this changer we update the CLI size-tracking changes for uncompressed
main-es2015 file. This file is larger due to new emitted shape of
ES2015 classes in TypeScript 3.9, which are now wrapped in IIFE.

alan-agius4 committed:
test: enable importHelpers for UMD builds

This is a workaround for a TS 3.9 regression https://github.com/microsoft/TypeScript/issues/38501 where the emitted `__exportStar` helpers have a missing semi-colon at the end of the unnamed function, when targetting UMD, and causes the following runtime error `Uncaught TypeError: (intermediate value)(…) is not a function`.

This is because the anonymous `__exportStar` function will be invoked with the function on the next like as the parameter which is subsequently invoking whatever was returned.

To get around this TS bug, add `importHelpers: true` in your tsconfig. This also, is recommanded to avoid multiple copies of the same helper being inlined, which might cause increase in bundle size.

alan-agius4 commented:
@googlebot I consent

alan-agius4 committed:
test: disable `cli-hello-world-lazy-rollup` from ivy tests

`cli-hello-world-lazy-rollup` fails on a bundle size check because Ivy and VE main-es2015 sizes are different

atscott commented on packages/compiler-cli/src/typescript_support.ts:
```diff
@@ -19,7 +19,7 @@ const MIN_TS_VERSION = '3.8.3';
  * MAX_TS_VERSION is not considered as a supported TypeScript version
  */
-const MAX_TS_VERSION = '3.9.0';
+const MAX_TS_VERSION = '3.10.0';
```
I think the next version is 4.0?

atscott commented on packages/compiler-cli/package.json:
```diff
@@ -27,7 +27,7 @@
     "@angular/compiler": "0.0.0-PLACEHOLDER",
     "tslib": "^1.10.0",
-    "typescript": ">=3.8 <3.9"
+    "typescript": ">=3.8 <3.10"
```
Same here, I think the next one is 4.0, not 3.10. https://www.npmjs.com/package/typescript

atscott commented on packages/bazel/package.json:
```diff
@@ -26,15 +26,15 @@
     "@bazel/typescript": ">=1.0.0",
     "terser": "^4.3.1",
-    "typescript": ">=3.8 <3.9",
+    "typescript": ">=3.8 <3.10",
```
4.0

atscott reviewed (approved):
Closing https://github.com/angular/angular/pull/37056 in favor of this one.

AndrewKushnir reviewed (approved):
LGTM 👍 

Reviewed-for: fw-forms

IgorMinar reviewed (approved):
LGTM! Thank you!

(I know that this was a painful PR so I'm optimizing for getting the PR in with as few additional changes as possible and creating JIRAs for the pending work.)

Reviewed-for: global-approvers

I have not reviewed in depth for the ngcc changes, I'll rely on other reviewers for that.

IgorMinar commented on goldens/size-tracking/integration-payloads.json:
```diff
@@ -60,7 +60,8 @@
         "bundle": "TODO(i): we should define ngDevMode to false in Closure, but --define only works in the global scope.",
-        "bundle": 169991
+        "bundle": "TODO(i): TS 3.9 new class shape seems to have broken Closure in big ways. The size went from 169991 to 252338",
+        "bundle": 252338
```
Can you please create v10.0 jira for this (feel free to copy the info from above) and reference the jira from this TODO? thanks?

alan-agius4 replied:
FW-2164

alxhub commented on packages/compiler-cli/ngcc/src/locking/lock_file_with_child_process/index.ts:
```diff
@@ -78,7 +78,7 @@ export class LockFileWithChildProcess implements LockFile {
     const isWindows = process.platform === 'win32';
     const unlocker = fork(
-        this.fs.resolve(__dirname, './unlocker.js'), [path, logLevel],
+        __dirname + '/unlocker.js', [path, logLevel],
```
Alternatively, we could replace `this.fs.resolve` with a call to the real `path.resolve`.

petebacondarwin replied:
`__dirname` is guaranteed to be absolute already, so there is no benefit in calling `resolve()` and is just unnecessary processing.

alxhub commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -602,6 +553,186 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+   * @returns the symbol for the class or `undefined` if `declaration` does not represent an outer
+   *     declaration of a class.
+   */
+  protected getClassSymbolFromOuterDeclaration(declaration: ts.Node): NgccClassSymbol|undefined {
```
Prefer `null` to `undefined`.

petebacondarwin replied:
The thing is that this method is only called by `getClassSymbol()`, which also returns `undefined` and is part of the `NgccReflectionHost` interface. So either one would need to change that interface to return `null`, which I would argue is outside the scope of this PR; or we would return `null` here and then convert it back to `undefined` in the implementation of `getClassSymbol()`, which seems pointless.

alxhub commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
````diff
@@ -892,6 +1023,33 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+    // ```
+    if (symbol.adjacent !== undefined) {
+      if (ts.isVariableDeclaration(symbol.adjacent.valueDeclaration)) {
+        if (symbol.adjacent.exports) {
````
Don't rely on truthiness, specifically check `!== undefined`.

alxhub commented on packages/core/src/event_emitter.ts:
```diff
@@ -61,7 +61,7 @@ import {Subject, Subscription} from 'rxjs';
  * @publicApi
  */
-export class EventEmitter<T extends any> extends Subject<T> {
+export class EventEmitter<T = any> extends Subject<T> {
```
This is very scary to me.

Lots of Angular code in the wild uses `EventEmitter`, and now we're allowing people to write `foo: EventEmitter` and get a default type parameter of `any`, whereas before providing one would be required. This is a major public API change.

IgorMinar replied:
You are right!!! I misread the diff. thanks for catching this!

We should not make this change as that would open up pandora's box for poorly typed code to be authored.

alxhub replied:
To be clear: one of the main problems with this is that template type-checking uses the generic type of `EventEmitter` to infer the type of `$event` in binding expressions for that event. Not requiring a generic parameter means that in cases where users don't provide it, `$event` will be `any`, which defeats the type-checking we try to do with it.

petebacondarwin replied:
I agree with @alxhub. I'm not sure what `T extends any any` provided over just `T` anyway.

alan-agius4 replied:
Essentially this change is because of the new ts change `Type Parameters That Extend any No Longer Act as any` where in this case `T extends any` acts like `T extends unknown`. This might case some breakages.

alan-agius4 replied:
I think seeing the other comments most likely this would be desired. So I’ll change this to `extends unknown`

petebacondarwin replied:
I think it should just be:

```suggestion
export class EventEmitter<T> extends Subject<T> {
```

alan-agius4 replied:
Both `EventEmitter<T>` and `EventEmitter<T extends any>`  will cause breakages because in this case` T` will act as `unknown`.

Also previously we did allow init the EventEmitter without a type https://github.com/angular/angular/blob/4f3ac1d983e93f5aaeb1bd9c719ada5644642df6/packages/forms/src/directives/ng_model.ts#L201, the problem now is that the return type would be `EventEmitter<unknown>` instead of  `EventEmitter<any>`.

Thanks to @petebacondarwin, we manage to find a solution to retain the same behaviour and not cause any breaking change.

alan-agius4 replied:
https://github.com/angular/angular/pull/36989/commits/3abaadb8b3fc426f47c603a67e8ba9fb8db13842

alan-agius4 commented on integration/BUILD.bazel:
```diff
@@ -54,7 +54,11 @@ INTEGRATION_TESTS = {
-    "cli-hello-world-lazy-rollup": {"commands": "payload_size_tracking"},
+    "cli-hello-world-lazy-rollup": {
+        "commands": "payload_size_tracking",
+        # TODO: cli-hello-world-lazy-rollup fails on a bundle size check because Ivy and VE sizes are different
```
FW-2165

alan-agius4 committed:
refactor: `EventEmitter` to retain behaviour of pre TypeScript 3.9

TypeScript 3.9 introduced a breaking change where extends `any` no longer acts as `any`, instead it acts as `unknown`.

With this change we retain the behavior we had with TS 3.8 which is;

When using the `EventEmitter` as a type you must always provide a  type;
```ts
let emitter: EventEmitter<string>
```

and when initializing the `EventEmitter` class you can either provide a  type or or use the fallback type which is `any`

```ts
const emitter = new EventEmitter(); // EventEmitter<any>
const emitter = new EventEmitte<string>(); // EventEmitter<string>
``

alan-agius4 committed:
test: update language service module resolution cache

With this change we update the expect of the `module resolution cache` were in the second count with cache the `fileExists` is called less then 700 times.

alan-agius4 commented:
**Note:** since TS3.9 is now final I dropped support for TS 3.8.

Maistho commented on packages/core/src/event_emitter.ts:
```diff
@@ -83,7 +90,7 @@ export class EventEmitter<T> extends Subject<T> {
    * @param value The value to emit.
    */
-  emit(value?: T) {
+  emit(value?: any) {
```
Would this also be considered a breaking change?

Previously, if I created an EventEmitter<string> it would enforce that I only emit strings, but with this change I would be able to emit anything. It will not cause any failures in existing code, but is a regression in the type-safety of the EventEmitter.

```ts
const emitter = EventEmitter<string>();
emitter.emit(123); // This would previously fail, but should succeed now.
```

mary-poppins commented:
You can preview 1302b4f at https://pr36989-1302b4f.ngbuilds.io/.
You can preview 50da797 at https://pr36989-50da797.ngbuilds.io/.
You can preview 8f771e2 at https://pr36989-8f771e2.ngbuilds.io/.
You can preview cbecf60 at https://pr36989-cbecf60.ngbuilds.io/.
You can preview 822bd51 at https://pr36989-822bd51.ngbuilds.io/.
You can preview 64ee66f at https://pr36989-64ee66f.ngbuilds.io/.
You can preview 845573e at https://pr36989-845573e.ngbuilds.io/.
You can preview 904c60f at https://pr36989-904c60f.ngbuilds.io/.
You can preview b66d67d at https://pr36989-b66d67d.ngbuilds.io/.
You can preview 43719a7 at https://pr36989-43719a7.ngbuilds.io/.
You can preview 172f66f at https://pr36989-172f66f.ngbuilds.io/.
You can preview 267c5a0 at https://pr36989-267c5a0.ngbuilds.io/.
You can preview bea7c09 at https://pr36989-bea7c09.ngbuilds.io/.
You can preview 3abaadb at https://pr36989-3abaadb.ngbuilds.io/.
You can preview 5b9ab36 at https://pr36989-5b9ab36.ngbuilds.io/.
You can preview c8aaf0c at https://pr36989-c8aaf0c.ngbuilds.io/.
You can preview fab602a at https://pr36989-fab602a.ngbuilds.io/.
You can preview 7007be8 at https://pr36989-7007be8.ngbuilds.io/.
You can preview 921a40f at https://pr36989-921a40f.ngbuilds.io/.
You can preview ec6d7a8 at https://pr36989-ec6d7a8.ngbuilds.io/.

petebacondarwin committed:
build(docs-infra): correctly handle "pseudo" classes

In the code base there are cases where there is, conceptually, a class
that is represented by a combination of an `interface`
(type declaration) and a `const` (value declaration).

For example:

```
export interface SomeClass {
  count(a?: string): number;
}
export const: SomeClass = class {
  someMethod(a: string = ''): number { ... }
};
```

These were being rendered as interfaces and also not
correctly showing the descriptions and default parameter
values.

In this commit such concepts are now rendered as classes.
The classes that are affected by this are:

* `DebugElement`
* `DebugNode`
* `Type`
* `EventEmitter`
* `TestBed`

Note that while decorators are also defined in this form
they have their own rendering type (`decorator`) and so
are not affecte by this.

mary-poppins commented:
You can preview 9baf51b at https://pr36989-9baf51b.ngbuilds.io/.

kara commented:
[presubmit](http://test/OCL:311360974:BASE:311361022:1589392252245:e1c4c92f)

petebacondarwin commented on packages/compiler-cli/ngcc/src/rendering/esm_rendering_formatter.ts:
```diff
@@ -104,7 +104,8 @@ export class EsmRenderingFormatter implements RenderingFormatter {
     }
-    const declarationStatement = getDeclarationStatement(classSymbol.declaration.valueDeclaration);
+    const declarationStatement =
+        getContainingStatement(classSymbol.implementation.valueDeclaration);
```
If there is no IIFE then it is the same thing.

petebacondarwin commented on packages/compiler-cli/test/transformers/node_emitter_spec.ts:
```diff
@@ -576,6 +592,10 @@ function normalizeResult(result: string, format: Format): string {
 
+  // Remove hoisted initial export assignments. These were added in TS 3.9:
+  // https://github.com/Microsoft/TypeScript/commit/c6c2c4c8d5aa0947de16f484b8c16fb0eab1c48f
+  res = res.replace(/^exports\.\w+ = void 0;$/gm, '');
```
the `\w+` needs to match any valid javascript identifier

gkalpak replied:
In this case, I think it is safe to use `/^exports\.\S+ = void 0;$/gm`.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/commonjs_umd_utils.ts:
```diff
@@ -87,8 +89,8 @@ export function isExportStatement(stmt: ts.Statement): stmt is ExportStatement {
- * Check whether the specified `ts.Statement` is a re-export statement, i.e. an expression statement
- * of one of the following forms:
+ * Check whether the specified `ts.Statement` is a wildcard re-export statement.
+ * I.E. an expression statement of one of the following forms:
```
I never know !

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/commonjs_umd_utils.ts:
```diff
@@ -56,7 +56,7 @@ export interface RequireCall extends ts.CallExpression {
  */
 export function findNamespaceOfIdentifier(id: ts.Identifier): ts.Identifier|null {
-  return id.parent && ts.isPropertyAccessExpression(id.parent) &&
+  return id.parent && ts.isPropertyAccessExpression(id.parent) && id.parent.name === id &&
```
Because we have a situation where the parent has

```
{
  name: X
  expression: Y
}
```

And we only want to check that `id` is the `name` property and it is an identifier.
Otherwise we could pass in the `expression` property as `id` and it would still return true.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/commonjs_host.ts:
```diff
@@ -152,18 +139,13 @@ export class CommonJsReflectionHost extends Esm5ReflectionHost {
     }
 
-    const viaModule = stripExtension(importedFile.fileName);
+    const viaModule = importPath.startsWith('./') ? null : importPath;
```
Good spot! Throughout the codebase it has `startsWith('.')` but I wanted to avoid the case where the file begins with a `.`. But you are right this is too restrictive.

petebacondarwin replied:
Agh! I changed the UMD one and not this...

gkalpak replied:
Not that importing local files (even ones starting with a `.`) would still need to be prefixed with `.`. The question is whether there can be npm packages whose name starts with `.`.

And according to [this doc][1], packages are not allowed to start with a dot:
>The name can’t start with a dot or an underscore.

So, `startsWith('.')` should be enough (but, of course, being more explicit won't hurt either).

[1]: https://docs.npmjs.com/files/package.json#name

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/commonjs_umd_utils.ts:
```diff
@@ -133,6 +147,69 @@ export function isWildcardReexportStatement(stmt: ts.Statement): stmt is Wildcar
+  }
+
+  // Check for Object.defineProperty
+  if (!ts.isPropertyAccessExpression(stmt.expression.expression) ||
+      !ts.isIdentifier(stmt.expression.expression.expression) ||
+      stmt.expression.expression.expression.text !== 'Object' ||
+      !ts.isIdentifier(stmt.expression.expression.name) ||
```
Yes, let's follow up with a refactor PR.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/commonjs_umd_utils.ts:
```diff
@@ -133,6 +147,69 @@ export function isWildcardReexportStatement(stmt: ts.Statement): stmt is Wildcar
+    return false;
+  }
+
+  const args = stmt.expression.arguments;
+  if (args.length !== 3) {
+    return false;
+  }
+  const exportsObject = args[0];
+  if (!ts.isIdentifier(exportsObject) || exportsObject.text !== 'exports') {
+    return false;
+  }
+
+  const propertyKey = args[1];
+  if (!ts.isStringLiteral(propertyKey)) {
+    return false;
+  }
+
+  const propertyDescriptor = args[2];
+  if (!ts.isObjectLiteralExpression(propertyDescriptor)) {
+    return false;
+  }
+
+  return (propertyDescriptor.properties.some(
+      prop => prop.name !== undefined && ts.isIdentifier(prop.name) && prop.name.text === 'get'));
```
I believe there are integration tests that will fail if this doesn't work. But I can't remember off the top of my head. Let's circle back on this in a subsequent PR if necessary.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -90,13 +91,18 @@ export class UmdReflectionHost extends Esm5ReflectionHost {
     for (const statement of this.getModuleStatements(sourceFile)) {
       if (isExportStatement(statement)) {
-        const exportDeclaration = this.extractUmdExportDeclaration(statement);
+        const exportDeclaration = this.extractBasicUmdExportDeclaration(statement);
```
😱 🙀 😱 🙀 😱 🙀 😱 🙀

IgorMinar reviewed (approved):
LGTM again - now that public API has changed a bit - thanks @petebacondarwin 

Reviewed-for: public-api

petebacondarwin commented on modules/benchmarks/src/util.ts:
```diff
@@ -41,7 +41,8 @@ export function bindAction(selector: string, callback: () => void) {
 export function profile(create: () => void, destroy: () => void, name: string) {
   return function() {
-    window.console.profile(name);
+    // 'cosole.profile' is experimental and was removed from DOM lib in TS 3.9
```
Ugh

petebacondarwin commented on modules/benchmarks/src/util.ts:
```diff
@@ -52,7 +53,8 @@ export function profile(create: () => void, destroy: () => void, name: string) {
       destroy();
     }
-    window.console.profileEnd();
+    // 'cosole.profileEnd' is experimental and was removed from DOM lib in TS 3.9
```
Ugh2

mary-poppins commented:
You can preview 7f29cec at https://pr36989-7f29cec.ngbuilds.io/.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -202,25 +217,67 @@ export class UmdReflectionHost extends Esm5ReflectionHost {
     }
 
-    // We need to add the `viaModule` because  the `getExportsOfModule()` call
+    // We need to compute the `viaModule` because  the `getExportsOfModule()` call
     // did not know that we were importing the declaration.
```
The two lines above should be removed.

mary-poppins commented:
You can preview 049cbfc at https://pr36989-049cbfc.ngbuilds.io/.

petebacondarwin commented on packages/language-service/test/reflector_host_spec.ts:
```diff
@@ -59,8 +59,8 @@ describe('reflector_host_spec', () => {
-    expect(secondCount).toBeGreaterThan(700);
-    expect(secondCount).toBeLessThan(800);
+    expect(secondCount).toBeGreaterThan(500);
+    expect(secondCount).toBeLessThan(600);
```
@kyliau - it would be good to understand why these have changed...

alan-agius4 replied:
Aperently, in 3.9 TS improved file lookups.

alan-agius4 replied:
What is strange though that the lookups decreased after the event emitter changes.

petebacondarwin commented on goldens/size-tracking/aio-payloads.json:
```diff
@@ -12,7 +12,7 @@
     "master": {
```
JIRA created: FW-2166

mhevery commented:
[presubmit](http://test/OCL:311402507:BASE:311402549:1589404662720:737b198a)

mary-poppins commented:
You can preview 041cec5 at https://pr36989-041cec5.ngbuilds.io/.

IgorMinar commented:
@petebacondarwin your suggestion worked and the code is now fully compatible with closure! yay! awesome job!

I resolved the merge conflict that occured due to recent merges upstream, updated the payload sizes for aio-local, rebased the branch, force-pushed, and kicked of a g3sync presubmit. 

Hopefully this will make the PR go green 🤞

mary-poppins commented:
You can preview d53f40e at https://pr36989-d53f40e.ngbuilds.io/.

IgorMinar commented:
[presubmit](https://test.corp.google.com/ui#id=OCL:311175159:BASE:311438878:1589416803897:1df02f88)

mary-poppins commented:
You can preview 11cf068 at https://pr36989-11cf068.ngbuilds.io/.
